### PR TITLE
feat: lance ray data sink

### PR DESCRIPTION
**Pull request by eddyxu**
Provide a `LanceDatasink` to be used with `ray.data`.

It offers distributed write and compute via Ray

### CODE REVIEW

File: python/python/lance/ray/sink.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+
+        self.uri = uri
+        self.schema = schema
+        self.mode = mode
+
+        if self.mode in set(["create", "overwrite"]) and self.schema is None:
+            raise ValueError("schema is required for create and overwrite modes")
```
Comment (westonpace) [inline python/python/lance/ray/sink.py:77]: Not a big deal but is there any way we can determine the schema from the first batch of data?

File: python/python/lance/ray/sink.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    Write a Ray dataset to lance.
+
+    Parameters
+    ----------
+    uri : str
+        The base URI of the dataset.
+    schema : pyarrow.Schema
+        The schema of the dataset.
+    mode : str, optional
+        The write mode. Default is 'append'. Choices are 'append' and 'overwrite'.
```
Comment (westonpace) [inline python/python/lance/ray/sink.py:32]: Looks like the default is create which isn't even mentioned here?

File: python/python/lance/ray/sink.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+        self.uri = uri
+        self.schema = schema
+        self.mode = mode
+
+        if self.mode in set(["create", "overwrite"]) and self.schema is None:
+            raise ValueError("schema is required for create and overwrite modes")
+
+        self.max_rows_per_file = max_rows_per_file
+        self.read_version: int | None = None
```
Comment (westonpace) [inline python/python/lance/ray/sink.py:80]: This isn't documented.

**Review by westonpace — APPROVED** [review]
A few minor suggestions but looks cool

File: python/python/lance/ray/sink.py (hunk 1 above)
Comment (eddyxu) [reply python/python/lance/ray/sink.py]: Add a `Literal[]` for type annotation.

File: python/python/lance/ray/sink.py (hunk 2 above)
Comment (eddyxu) [reply python/python/lance/ray/sink.py]: Fixed.

File: python/python/lance/ray/sink.py (hunk 3 above)
Comment (eddyxu) [reply python/python/lance/ray/sink.py:80]: Add a docstring. This is internal behavior (for now).

**Comment by zeddit** [comment]
looking forward to this feature. 
does this feature support writing dataset to s3 backend, and has it benchmarked the write performance and guarantee the insert order for each worker. great thanks.

**Pull request merged**